lua/entities/gmod_wire_expression2/core/cl_http_cl.lua

```diff
@@ -0,0 +1,131 @@
+--[[
+
+	Clientside HTTP library
+	 - swadical 🍌
+
+]]
+
+--	Clientside HTTP is separate from serverside HTTP:
+--		Serverside HTTP allows greater speeds than clientside HTTP and it is
+--		up to the server owner to decide which extension should be enabled
+--		through a convar (found in core/http.lua)
+
+--	This library will mirror the functionality of core/http.lua as best as
+--	possible, making sure any existing e2s which make use of the http library
+--	are not broken by this extension
+
+E2Lib.clHTTP = {}
+local lib = E2Lib.clHTTP
+
+lib.netMsgID = "wire_expression2_cl_http" -- open to change
+lib.currentUID = 0
+
+-- number function(void)
+function lib:newUID()
+	self.currentUID = (self.currentUID + 1) % 255
+	return self.currentUID -- return a number between 0 and 254
+end
+
+lib.rawRequest = http.Fetch
+lib.segmentSize = 4 * 1024-- 4 kilobytes
+lib.segmentBits = math.min(math.ceil(math.log(lib.segmentSize,2)) + 2,32)
+lib.sendInterval = 0.5-- seconds
+lib.HTTP_REQUEST_FAILED = 1024--enum for HTTP failure
+
+-- void function(string id, number interval, function func)
+-- Internal: Do not call.
+function lib.timerCoroutine(id,interval,func)
+	func = coroutine.wrap(func)
+	timer.Create(id,interval,0,function()
+		if func() then
+			timer.Remove(id)
+		end
+	end)
+end
+
+-- string, string function(void)
+-- Internal: Do not call.
+function lib:decodeRequest()
+	return net.ReadString(),net.ReadUInt(8)
+end
+
+-- void function(void)
+-- Internal: Do not call.
+function lib:handleIncomingRequest()
+	local url,uid = self:decodeRequest()
+
+	self:performRequest(url,uid)
+end
+
+-- void function(string url, number uid)
+-- Internal: Do not call.
+function lib:performRequest(url,uid)
+	self.rawRequest(url,function(body,length,headers,code)
+		self.timerCoroutine("wire_e2_cl_http_"..uid,self.sendInterval,function()
+			return self:returnData(uid,body,length,headers,code)
+		end)
+	end,function(err)
+		self:returnFailure(uid,err)
+	end)
+end
+
+-- void function(number uid, string body, number length, table headers, number code)
+-- Internal: Do not call.
+function lib:returnData(uid,body,length,headers,code)
+	local body_compressed = util.Compress(body) or ""
+
+	self:writeHTTPBody(uid,body_compressed,#body_compressed)
+	self:writeMetadata(uid,length,headers,code)
```

Comment:
The drawback of doing it this way is that you have to send a success/error boolean in every HTTPBody chunk. The advantage is that you don't have to send the number of chunks and/or a "I'm finished" message at the end.